Speedup

altaite opened the pull request:
Several changes in decoder, resulting in about 4x speedup of BioJava MMTF parsing. Largest change (but not most significant in terms of time) is replacing Jackson with Jmol version of MessagePack parsing (sources were extracted, no dependency). A test for that was created, comparing objects created by both methods.

altaite committed:
Modified access to resource files so that it works on Windows and 
hopefully every other OS.

altaite committed:
Replaced Jackson by MessagePack parsing code from Jmol. Jackson remains
available by a switch in MessagePackSerialization.

altaite committed:
Moving code from getNumBonds() to constructor results in 4-fold speedup
of BioJava Structure creation.

altaite committed:
FLOAT64 is now parsed as double, which corresponds to what Jackson did.
Passed MessagePackTest.

altaite committed:
Added a test comparing if results of Jackson and the new MessagePack
parser adopted from Jmol are identical.

altaite committed:
Moved MessagePack test into mmtf-codec.

altaite committed:
Version set to 1.05.

josemduarte commented:
Thanks. I'm trying to review it, give me some time. As an advice for next time, please don't do code reformatting at the same time that you do important changes. The diffs as they are shown are not very helpful.

Having said that, there's a really nice trick to ignore whitespaces in the diff that will make them more readable: add a `?w=1` at the end of the URL (see https://github.com/blog/967-github-secrets). For this PR's diff it would be: https://github.com/rcsb/mmtf-java/pull/30/files?w=1

josemduarte commented on mmtf-codec/src/test/java/org/rcsb/mmtf/encoder/TestEncoderUtils.java:
```diff
@@ -1,169 +1,169 @@
+			return new GenericDecoder(ReaderUtils.getDataFromFile(inFile));
+		} catch (IOException e) {
+			e.printStackTrace();
+			throw new RuntimeException();
```
This is not your change, but it'd be nice to fix: don't catch the exception, just declare a "throws IOException" and error handling will be much clearer.

altaite replied:
Text files are in gz now.

josemduarte commented on mmtf-api/pom.xml:
```diff
@@ -1,54 +1,54 @@
+	<parent>
+		<groupId>org.rcsb</groupId>
+		<artifactId>mmtf</artifactId>
+		<version>1.0.5-SNAPSHOT</version>
```
No need to change the version here, SNAPSHOT versions are for development. Versions in pom should only change when a release happens. Same applies to other pom files.

josemduarte reviewed (changes requested):
It looks to me like the file `pdb_entry_type.txt` shouldn't be here, it's quite a large file. Is it needed?

josemduarte commented on mmtf-codec/src/test/java/org/rcsb/mmtf/serialization/mp/LineFile.java:
```diff
@@ -0,0 +1,41 @@
+				lines.add(line);
+			}
+			return lines;
+		} catch (IOException e) {
```
Better to declare the exception as thrown in the method signature. Error handling will be easier like that

josemduarte commented on mmtf-codec/src/test/java/org/rcsb/mmtf/serialization/mp/MessagePackTest.java:
```diff
@@ -0,0 +1,123 @@
+			}
+			return codes;
+		} catch (Exception ex) {
+			throw new RuntimeException(ex);
```
Just declare the exception as `throws` in signature. It will have the same effect and will be clearer

josemduarte commented on mmtf-codec/src/test/java/org/rcsb/mmtf/serialization/mp/MessagePackTest.java:
```diff
@@ -0,0 +1,123 @@
+			MessagePackSerialization.setJackson(true);
+			StructureDataInterface sdiJackson = parse(bytes);
+
+			ReflectionAssert.assertReflectionEquals(sdiJackson, sdiJmol);
```
Is there a reason to use this assert method instead of the standard assert from the junit library?

altaite replied:
This assert performs deep comparison, while standard assert just checks boolean.

josemduarte commented on mmtf-codec/src/test/java/org/rcsb/mmtf/serialization/mp/MessagePackTest.java:
```diff
@@ -0,0 +1,123 @@
+ * creation gives identical data.
+ *
+ */
+public class MessagePackTest extends TestCase {
```
Better use junit4. It doesn't require extending TestCase, just adding the \@Test annotation to the method you want to make into a test

josemduarte commented on mmtf-serialization/pom.xml:
```diff
@@ -1,296 +1,93 @@
+            <groupId>org.unitils</groupId>
+            <artifactId>unitils-core</artifactId>
+            <version>3.4.2</version>
+            <scope>test</scope>
```
Is this really needed? Can't we just use junit4?

altaite replied:
Yes, for the deep comparison of objects.

josemduarte commented on mmtf-serialization/src/main/java/org/rcsb/mmtf/serialization/MessagePackSerialization.java:
```diff
@@ -1,51 +1,92 @@
+			mmtfBean = objectMapper.readValue(inputStream, MmtfStructure.class);
+		} catch (IOException e) {
+			e.printStackTrace();
+		}
```
Exception should be declared in signature and not caught

josemduarte commented on mmtf-serialization/src/main/java/org/rcsb/mmtf/serialization/MessagePackSerialization.java:
```diff
@@ -1,51 +1,92 @@
+			return s;
+		} catch (Exception e) {
+			throw new RuntimeException(e);
+		}
```
Declare IOException in signature

josemduarte commented on mmtf-serialization/src/main/java/org/rcsb/mmtf/serialization/MessagePackSerialization.java:
```diff
@@ -1,51 +1,92 @@
+		try {
+			objectMapper.writeValue(outputStream, mmtfStructure);
+		} catch (IOException e) {
+			e.printStackTrace();
```
Don't catch the exception, declare it in signature

josemduarte commented:
One concern is the 17.9% drop in test coverage, I guess because of the new message pack decoding code from jmol that has no tests. Do you have some metrics on the performance difference of the new decoder versus the msgpack library? It's a pity to lose a well-tested library in favor of an ad-hoc decoder.

Anyway thanks a lot for the patch! looks good in general. If you could address the comments that'd be great.

altaite committed:
Renaming of mp package to quickmessagepackdeserialization.

altaite committed:
Zipped resource pdb_entry_type.txt. Replaced few other
RuntimeExceptions by IOException and try with resources.

josemduarte commented on mmtf-codec/src/test/java/org/rcsb/mmtf/serialization/mp/ZippedTextFile.java:
```diff
@@ -0,0 +1,29 @@
+		}
+	}
+
+}
```
You need gzipping rather than zipping here.

This is how you do it with gzipping:
```java
public static void gzipFile(File inFile, File outFile) throws IOException {
		GZIPOutputStream zos = new GZIPOutputStream(new FileOutputStream(outFile));
		FileInputStream is = new FileInputStream(inFile);

		int b;
		while ( (b=is.read())!=-1) {
			zos.write(b);
		}
		zos.close();
		is.close();
	}
```
And ungzip is:
```java
public static void gunzipFile(File inFile, File outFile) throws IOException {
		GZIPInputStream zis = new GZIPInputStream(new FileInputStream(inFile));
		FileOutputStream os = new FileOutputStream(outFile);
		int b;
		while ( (b=zis.read())!=-1) {
			os.write(b);
		}
		zis.close();
		os.close();
	}
```

altaite replied:
I would say that my code is probably faster (buffer). And more importantly, it is easier to prepare the resource, because I do not need to do it with Java. I just do it by file manager or command line, which is far more convenient. Also I would leave it because it does not matter much I think.

josemduarte replied:
The issue here is zip vs gzip, not so much about efficiency. Zip is an archive format (to put a bunch of files together), while gzip is purely a compression format. 

You can also do gzip easily from command line with tools like gzip and gunzip, so that's also not a concern.

altaite committed:
Replaced zip with gz. Also corrected the loading of file using URL,
previous was failing on Windows again. Going to use URL elsewhere 
too in the next commit.

altaite committed:
Nicer code for reading text files as resources in gz. URL are used
whenever possible (simpler and more reliable), File left only when 
necessary for API tests.

altaite committed:
Minor formatting changes and removal of dead code in comments.

altaite commented:
Concerning test coverage (I do not see way how to comment directly):
There is a test which allows to compare the new code creates exactly the same objects as Jackson, on the whole PDB: MessagePackTest. But it must be in different subproject mmtf-codec because of dependencies. Is there some easy way how to connect this to the code, so that coverage would be better? It is integration test, but it test equality of Jackson and new code completely, if you run it on whole database.

josemduarte commented:
Thanks for all the changes! the only thing missing is putting back the version to 1.0.4-SNAPSHOT. After that I think we should merge. By the way there's a merge conflict now, could you take care of that too? (just merge into this branch the latest master). The conflict should be minimal,  the only file that diverged since you branched is `mmtf-codec/src/main/java/org/rcsb/mmtf/decoder/ReaderUtils.java`

josemduarte commented on mmtf-serialization/src/main/java/org/rcsb/mmtf/serialization/quickmessagepackdeserialization/SB.java:
```diff
@@ -0,0 +1,354 @@
+ * 
+ */
+
+public class SB {
```
Just found this one: could we get rid of it? This looks like a replacement for StringBuilder needed in jmol so that the java to js transpiler is happy. We don't need it here, let's replace it by StringBuilder

josemduarte commented on mmtf-serialization/src/main/java/org/rcsb/mmtf/serialization/quickmessagepackdeserialization/ObjectTree.java:
```diff
@@ -0,0 +1,124 @@
+ *
+ * @author Antonin Pavelka
+ */
+public class ObjectTree {
```
Could we have some documentation for this class? or at least change the method names to something more usable, one letter method names without javadocs are going to be difficult to understand....

josemduarte commented on mmtf-serialization/src/main/java/org/rcsb/mmtf/serialization/quickmessagepackdeserialization/ObjectTree.java:
```diff
@@ -0,0 +1,124 @@
+		}
+	}
+
+	public Hashtable l(String s) {
```
This method doesn't seem to be used. Do we need it?

josemduarte commented on mmtf-serialization/src/main/java/org/rcsb/mmtf/dataholders/MmtfStructureFactory.java:
```diff
@@ -0,0 +1,128 @@
+ *
+ * @author Antonin Pavelka
+ */
+public class MmtfStructureFactory {
```
Could you add some javadocs to explain what this class does? The individual methods are more self-explanatory

altaite committed:
Message Pack parsing code from Bob Hanson's Jmol simplified and moved to one class.

altaite committed:
PDB codes for testing now does not contain missing records (no codes after 12/01/2016). Increased -Xmx for
testing from 1500 to 6000 MB because sometimes GC timeout occured in MessagePackTest. Improved comments on the MessagePack parser.

altaite committed:
Correction of versions to 1.0.4-SNAPSHOT.

altaite committed:
Better API for Lines utility.

altaite committed:
Merge branch 'master' into speedup

altaite committed:
Better API for Lines utility. Also number of tested structures
was corrected to reasonable number 10.

altaite committed:
Merge origin/speedup into speedup

altaite committed:
Replaced Exception for ParseException in MessagePack decoding.

altaite committed:
More appropriate name for test.

josemduarte reviewed (changes requested):
This looks really good now, thank you so much for all the changes. Just one final detail: I'd propose method names for `ObjectTree` like `getString`, `getInteger`, `getByteArray` etc. As I said, single letter names make things difficult to understand. Let's merge after you refactor those, it should only take a couple of minutes.

altaite committed:
More descriptive method names in ObjectTree.
